src/dask_awkward/lib/core.py

```diff
@@ -1530,22 +1564,43 @@ def map_partitions(
     This is effectively the same as `d = c * a`
 
     """
-    token = token or tokenize(fn, *args, meta, **kwargs)
-    label = label or funcname(fn)
+    token = token or tokenize(base_fn, *args, meta, **kwargs)
+    label = label or funcname(base_fn)
     name = f"{label}-{token}"
+    kwarg_deps, kwarg_repacker = unpack_collections(kwargs, traverse=traverse)
+    deps, _ = unpack_collections(*args, *kwargs.values(), traverse=traverse)
+
+    arg_deps_expanded = []
+    arg_repackers = []
+    arg_lens_for_repackers = []
+    for arg in args:
+        this_arg_deps, repacker = unpack_collections(arg, traverse=traverse)
+        if len(this_arg_deps) > 0:
+            arg_deps_expanded.extend(this_arg_deps)
+            arg_repackers.append(repacker)
+            arg_lens_for_repackers.append(len(this_arg_deps))
+        else:
+            arg_deps_expanded.append(arg)
+            arg_repackers.append(None)
+            arg_lens_for_repackers.append(1)
+
+    fn = ArgsKwargsPackedFunction(
+        base_fn,
+        arg_repackers,
+        kwarg_repacker,
+        arg_lens_for_repackers,
+    )
+
     lay = partitionwise_layer(
         fn,
         name,
-        *args,
+        *arg_deps_expanded,
+        *kwarg_deps,
         opt_touch_all=opt_touch_all,
-        **kwargs,
     )
-    deps = [a for a in args if is_dask_collection(a)] + [
-        v for _, v in kwargs.items() if is_dask_collection(v)
-    ]
 
     if meta is None:
-        meta = map_meta(fn, *args, **kwargs)
+        meta = map_meta(fn, *arg_deps_expanded, *kwarg_deps)
```

Comment:
if the function expects or requires named kwargs does `*kwarg_deps` not expand enough information (that is, why don't we need dict `**` unrolling here?). It looks like tests are working so it seems this is indeed correct! I'm just wondering how

Reply:
`kwarg_deps` is guaranteed to be a un-nested list of unlabeled arguments (the first return value of `unpack_collections`). It's the repack function that carries around the actual args/kwargs structure, if there is any.

Reply:
Yeah it's a bit odd that kwargs_deps is actually a flat list of arguments that need to be repacked into kwargs.
I'll figure out a better name.